src/Lorentz/Contracts/BaseDAO/Types.hs

```diff
@@ -40,73 +153,201 @@ instance TypeHasDoc MigrationStatus where
+  , sProposalKeyListSortByDate :: [ProposalKey] -- Newest first (probably change to set)
   }
   deriving stock (Generic, Show)
-  deriving anyclass (IsoValue, HasAnnotation)
+  deriving anyclass (HasAnnotation)
 
-instance TypeHasDoc Storage where
-  typeDocMdDescription =
-    "Contract's storage holding a big_map with all balances and the operators."
+deriving anyclass instance (WellTypedIsoValue proposalMetadata) => IsoValue (Storage proposalMetadata)
 
-instance HasFieldOfType Storage name field =>
-         StoreHasField Storage name field where
+instance HasFieldOfType (Storage pm) name field =>
+         StoreHasField (Storage pm) name field where
   storeFieldOps = storeFieldOpsADT
 
-type StorageC store =
-  StorageContains store
-    [ "sAdmin" := Address
+instance IsoValue pm =>
+    StoreHasSubmap (Storage pm) "sLedger" LedgerKey LedgerValue where
+  storeSubmapOps = storeSubmapOpsDeeper #sLedger
+
+instance IsoValue pm =>
+    StoreHasSubmap (Storage pm) "sOperators" (Address, Address) () where
+  storeSubmapOps = storeSubmapOpsDeeper #sOperators
+
+instance IsoValue pm =>
+    StoreHasSubmap (Storage pm) "sProposals" ProposalKey (Proposal pm) where
+  storeSubmapOps = storeSubmapOpsDeeper #sProposals
+
+type StorageC store pm =
+  ( StorageContains store
+    [ "sLedger" := LedgerKey ~> LedgerValue
+    , "sOperators" := (Address, Address) ~> ()
+    , "sTokenAddress" := Address
+    , "sAdmin" := Address
     , "sPendingOwner" := Maybe Address
     , "sLedger" := Ledger
     , "sMigrationStatus" := MigrationStatus
+
+    , "sVotingPeriod" := VotingPeriod
+    , "sQuorumThreshold" := QuorumThreshold
+
+    , "sProposals" := ProposalKey ~> Proposal pm
+    , "sProposalKeyListSortByDate" := [ProposalKey]
     ]
+  )
 
 -- | Parameter of the BaseDAO contract
-data Parameter
+data Parameter proposalMetadata
   = Call_FA2 FA2.Parameter
   | Transfer_ownership TransferOwnershipParam
   | Accept_ownership ()
   | Migrate MigrateParam
   | Confirm_migration ()
+  | Propose (ProposeParams proposalMetadata)
+  | Proposal_metadata (View ProposalKey (Proposal proposalMetadata))
+  | Vote [VoteParam]
+  -- Admin
+  | Set_voting_period VotingPeriod
+  | Set_quorum_threshold QuorumThreshold
+  | Flush ()
   deriving stock (Generic, Show)
-  deriving anyclass (IsoValue)
 
-instance ParameterHasEntrypoints Parameter where
-  type ParameterEntrypointsDerivation Parameter = EpdDelegate
+instance (HasAnnotation pm, NiceParameter pm) => ParameterHasEntrypoints (Parameter pm) where
+  type ParameterEntrypointsDerivation (Parameter pm) = EpdDelegate
+
+deriving anyclass instance (WellTypedIsoValue pm) => IsoValue (Parameter pm)
 
 type TransferOwnershipParam = ("newOwner" :! Address)
 type MigrateParam = ("newAddress" :! Address)
 
-emptyStorage :: Storage
+-- | Voting period in seconds
+type VotingPeriod = Natural
+
+-- | QuorumThreshold that a proposal need to meet
+-- quorum_threshold = upvote - downvote
```

Comment:
I see at least 2 possible interpretations of quorum threshold. It can be `positive + negative` or `positive - negative`. I guess it can also be just `positive`. I am not sure which one is correct, I've asked [here]((https://tocquevillegroup.slack.com/archives/C01BTEPMY5U/p1603814934010800).

Reply:
UPD: so it's actually the total number of votes.

Reply:
I have refactored this according to our discussion from slack. Let me know if this is good enough, Ivan.